pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -403,42 +382,45 @@ func (c *MCRouteController) addMCFlowsForSingleCIImp(activeGW *mcv1alpha1.Gatewa
 	return nil
 }
 
-func (c *MCRouteController) deleteMCFlowsForSingleCIImp(ciImpName string) error {
+func (c *MCDefaultRouteController) deleteMCFlowsForSingleCIImp(ciImpName string) error {
 	if err := c.ofClient.UninstallMulticlusterFlows(ciImpName); err != nil {
 		return fmt.Errorf("failed to uninstall multi-cluster flows to remote Gateway Node %s: %v", ciImpName, err)
 	}
 	delete(c.installedCIImports, ciImpName)
 	return nil
 }
 
-func (c *MCRouteController) deleteMCFlowsForAllCIImps() error {
+func (c *MCDefaultRouteController) deleteMCFlowsForAllCIImps() error {
 	for _, ciImp := range c.installedCIImports {
 		c.deleteMCFlowsForSingleCIImp(ciImp.Name)
 	}
 	return nil
 }
 
-// getActiveGateway compares Gateway's CreationTimestamp to get the active Gateway,
-// The last created Gateway will be the active Gateway.
-func (c *MCRouteController) getActiveGateway() (*mcv1alpha1.Gateway, error) {
-	gws, err := c.gwLister.Gateways(c.namespace).List(labels.Everything())
+func (c *MCDefaultRouteController) getActiveGateway() (*mcv1alpha1.Gateway, error) {
+	activeGW, err := getActiveGateway(c.gwLister, c.namespace)
 	if err != nil {
 		return nil, err
 	}
-	if len(gws) == 0 {
+	if activeGW == nil {
 		return nil, nil
 	}
-	// Comparing Gateway's CreationTimestamp to get the last created Gateway.
-	lastCreatedGW := gws[0]
-	for _, gw := range gws {
-		if lastCreatedGW.CreationTimestamp.Before(&gw.CreationTimestamp) {
-			lastCreatedGW = gw
-		}
+	if net.ParseIP(activeGW.GatewayIP) == nil || net.ParseIP(activeGW.InternalIP) == nil {
+		return nil, fmt.Errorf("the active Gateway %s has no valid GatewayIP or InternalIP", activeGW.Name)
+	}
+	return activeGW, nil
+}
+
+func getActiveGateway(gwLister mclisters.GatewayLister, namespace string) (*mcv1alpha1.Gateway, error) {
+	gws, err := gwLister.Gateways(namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
 	}
-	if net.ParseIP(lastCreatedGW.GatewayIP) == nil || net.ParseIP(lastCreatedGW.InternalIP) == nil {
-		return nil, fmt.Errorf("the last created Gateway %s has no valid GatewayIP or InternalIP", lastCreatedGW.Name)
+	if len(gws) == 0 {
+		return nil, nil
 	}
-	return lastCreatedGW, nil
+	// The Gateway webhook guarantees there will be at most one Gateway in a cluster.
```

Comment:
Question - Gateway is created by MC Controller right? Do we still use a webhook to check only one Gateway? Because we assume users can also manually create Gateway and MC Controller does not remove redundant ones?

Reply:
Yes, we still need it. When there is already a Gateway created by MC controller, user won't be able to create a new Gateway because of the webhook. This is the original purpose to stop use to add it manually.

Reply:
I feel the MC Controller should reconcile and delete invalidate Gateways not reflecting the current state including the ones created by users, and that can cover the Controller restart case as well. Not really safe to assume webhook is always there. And in other code, better not to have that assumption either.

Reply:
I will check this part, probably a separate PR for this.